File: interview_kickstart/01_sorting_algorithms/class_discussed_problems/0252_meeting_rooms.js
/* 
First Intuition: 
  1. Sort intervals based on start time.
  2. Iterate through the array checking if start time at i + 1 is less
     than end time at i. Return false if it isnt.
  3. Return true after looping through all intervals.

Strategy: Presort

Time Complexity - O(nlogn) for sorting + O(n) to iterate and check
Space complexity - O(1) if mutating input + whatever space is being used
                  by js for sorting.
                  O(n) if not mutating input
*/

var canAttendMeetings = function (intervals) {
  const sorted = intervals.sort((a, b) => a[0] - b[0]),
    maxIndex = intervals.length - 1

  for (let i = 0; i < maxIndex; i++) {
    if (sorted[i][1] > sorted[i + 1][0]) return false
  }

  return true
}

// Tests
console.log(
  canAttendMeetings([
    [5, 10],
    [0, 30],
    [15, 20]
  ])
)
console.log(
  canAttendMeetings([
    [7, 10],
    [2, 4]
  ])
)

File: interview_kickstart/01_sorting_algorithms/class_discussed_problems/python/0252_meeting_rooms.py
from typing import List


class Solution:
    def canAttendMeetings(self, intervals: List[List[int]]) -> bool:
        intervals.sort()

        for i in range(len(intervals) - 1):
            if (intervals[i][1] > intervals[i+1][0]):
                return False

        return True


test = Solution()

print(test.canAttendMeetings([[0, 30], [5, 10], [15, 20]]))
print(test.canAttendMeetings([[7, 10], [2, 4]]))
